tests/flattenTest.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+/**
+ * @covers Dash\flatten
+ */
+class flattenTest extends PHPUnit_Framework_TestCase
+{
+	/**
+	 * @dataProvider cases
+	 */
+	public function test($iterable, $expected)
+	{
+		$this->assertSame($expected, Dash\flatten($iterable));
+	}
+
+	public function cases()
+	{
+		return [
+			'With nulls' => [
+				'iterable' => [null, null, null],
+				'expected' => [null, null, null],
+			],
+			'With empty iterable' => [
+				'iterable' => [],
+				'expected' => [],
+			],
+
+			/*
+				With indexed array
+			 */
+
+			'With an indexed array with only scalar values' => [
+				'iterable' => [
+					6,
+					1,
+					3,
+				],
+				'expected' => [6, 1, 3]
+			],
+			'With an indexed array with a mix of scalar and array values' => [
+				'iterable' => [
+					4,
+					[2, 3],
+					[7,9],
+					8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an indexed array with only array values' => [
+				'iterable' => [
+					[2, 3],
+					[7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With associative array
+			 */
+
+			'With an associative array with only scalar values' => [
+				'iterable' => [
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				],
+				'expected' => [ 6, 1, 3]
+			],
+			'With an associative array with a mix of scalar and array values' => [
+				'iterable' => [
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an associative array with only array values' => [
+				'iterable' => [
+					'b' => [2, 3],
+					'c' => [7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With stdClass
+			 */
+
+			'With an stdClass with only scalar values' => [
+				'iterable' => (object) [
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				],
+				'expected' => [6, 1, 3]
+			],
+			'With an stdClass with a mix of scalar and array values' => [
+				'iterable' => (object) [
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				],
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an stdClass with only array values' => [
+				'iterable' => (object) [
+					'b' => [2, 3],
+					'c' => [7,9],
+				],
+				'expected' => [2, 3, 7, 9]
+			],
+
+			/*
+				With ArrayObject
+			 */
+
+			'With an ArrayObject with only scalar values' => [
+				'iterable' => new ArrayObject([
+					'a' => 6,
+					'b' => 1,
+					'c' => 3,
+				]),
+				'expected' => [6, 1, 3]
+			],
+			'With an ArrayObject with a mix of scalar and array values' => [
+				'iterable' => new ArrayObject([
+					'a' => 4,
+					'b' => [2, 3],
+					'c' => [7,9],
+					'd' => 8,
+				]),
+				'expected' => [4, 2, 3, 7, 9, 8]
+			],
+			'With an ArrayObject with only array values' => [
+				'iterable' => new ArrayObject([
+					'b' => [2, 3],
+					'c' => [7,9],
+				]),
+				'expected' => [2, 3, 7, 9]
+			]
+		];
```

Comment:
Missing: test cases with arrays containing non-scalar/array data types, especially stdClass and `iterable` data types (eg. `ArrayObject`).

You might find that your implementation does funky things with `[stdClass, ...]`. Ah, PHP!

This also raises the question of whether iterable elements should be treated as arrays and unpacked, eg. `flatten([ new ArrayObject([1,2,3]) ]) == [1,2,3]`

Reply:
To answer the question raised in the previous comment: Throughout Dash we treat stdClass and other objects as iterable, but treating them as arrays here and unpacking their contents seems counterintuitive and unexpected.

My recommendation would be to only unpack arrays, and to force other iterables to be explicitly converted to array via a separate map() step.

Let me know if this comment is confusing and I’d be happy to explain more coherently in person.

Reply:
Let's discuss.  I don't follow.

Reply:
Will do. For now, here’s an example to illustrate: what should this produce?

    flatten([
        (object) [‘a’ => ‘one’, ‘b’ => ‘two’],
        (object) [‘c’ => ‘three’, ‘d’ => ‘four’]
    ])